tests/prague/eip7685_general_purpose_el_requests/test_deposits_withdrawals_consolidations.py

```diff
@@ -214,10 +215,140 @@ def get_contract_permutations(n: int = 3) -> Generator[Any, None, None]:
+                single_deposit_from_eoa(0),
+                single_consolidation_from_contract(0),
+                single_withdrawal_from_contract(0),
+            ],
+            id="deposit_from_eoa+consolidation_from_contract+withdrawal_from_contract",
+        ),
+        pytest.param(
+            [
+                single_consolidation_from_eoa(0),
+                single_consolidation_from_contract(1),
+                single_withdrawal_from_contract(0),
+            ],
+            id="consolidation_from_eoa+consolidation_from_contract+withdrawal_from_contract",
+        ),
+        # contract: consolidation + deposit
+        pytest.param(
+            [
+                single_withdrawal_from_eoa(0),
+                single_consolidation_from_contract(0),
+                single_deposit_from_contract(0),
+            ],
+            id="withdrawal_from_eoa+consolidation_from_contract+deposit_from_contract",
+        ),
+        pytest.param(
+            [
+                single_deposit_from_eoa(0),
+                single_consolidation_from_contract(0),
+                single_deposit_from_contract(1),
+            ],
+            id="deposit_from_eoa+consolidation_from_contract+deposit_from_contract",
+        ),
+        pytest.param(
+            [
+                single_consolidation_from_eoa(0),
+                single_consolidation_from_contract(1),
+                single_deposit_from_contract(0),
+            ],
+            id="consolidation_from_eoa+consolidation_from_contract+deposit_from_contract",
+        ),
+        # contract: withdrawal + deposit
+        pytest.param(
+            [
+                single_withdrawal_from_eoa(0),
+                single_withdrawal_from_contract(1),
+                single_deposit_from_contract(0),
+            ],
+            id="withdrawal_from_eoa+withdrawal_from_contract+deposit_from_contract",
+        ),
+        pytest.param(
+            [
+                single_deposit_from_eoa(0),
+                single_withdrawal_from_contract(0),
+                single_deposit_from_contract(1),
+            ],
+            id="deposit_from_eoa+withdrawal_from_contract+deposit_from_contract",
+        ),
+        pytest.param(
+            [
+                single_consolidation_from_eoa(0),
+                single_withdrawal_from_contract(0),
+                single_deposit_from_contract(0),
+            ],
+            id="consolidation_from_eoa+withdrawal_from_contract+deposit_from_contract",
+        ),
+        # testing upper limits of each request type per slot if it exists
+        pytest.param(
+            [
+                single_consolidation_from_contract(0),
+                single_consolidation_from_contract(1),
+                # single_consolidation_from_contract(2),  # i=2 is not allowed cuz
+                # only 2 MAX CONSOLIDATIONS PER BLOCK (EIP-7251), but why does the error message
+                # not inform about this? only reports hash mismatch
```

Comment:
That is because the limit not actually enforced by restricting the number of calls that can be made to the consolidation contract, rather, the consolidations executed after the limit is reached will queue up and only will be dequeued on the next block.

In general, the functions in this file that generate calls to the system contracts cannot know if their requests are going to be included in the block or not.

That's the reason behind the requests hash mismatch: the tx that generates the call is added to the block, it's executed, we add it to the list of expected requests (by our tests) in the header, but it will not match because of the queuing mechanism in the contract.

Hope I explained myself ok, but if not let me know 😄